File: Problems/01. Easy/Contains Duplicate/TypeScript/ContainsDuplicate.ts
function containsDuplicate(nums: number[]): boolean {
    // ...
    return false;
}

// Solution 1: Brute Force
function S1_containsDuplicate(nums: Array<number>): boolean {
    for (let i = 0; i < nums.length - 1; i++) {
        // O(n)
        let numToCompare = nums[i];
        for (let j = i + 1; j < nums.length; j++) {
            // O(n)
            let compareAgainst = nums[j];
            if (numToCompare === compareAgainst) return true;
        }
    }
    return false;
}

// Solution 2: Sort and Compare
function S2_containsDuplicate(nums: Array<number>): boolean {
    nums.sort((a, b) => a - b); // O(n log(n))
    for (let i = 0; i < nums.length - 1; i++) {
        // O(n)
        if (nums[i] === nums[i + 1]) {
            return true;
        }
    }
    return false;
}

// Solution 3: Set (Hash Table)
function S3_containsDuplicate(nums: Array<number>): boolean {
    const hash: Set<number> = new Set(); // O(n) space
    for (const num of nums) {
        // O(n) time
        if (hash.has(num)) {
            return true;
        }
        hash.add(num);
    }
    return false;
}

File: Problems/01. Easy/Contains Duplicate/Python/ContainsDuplicate.py
from typing import List

def contains_duplicate(nums):
    return
    

# Solution 1: Brute Force
def S1_contains_duplicate(nums: List[int]) -> bool:
    for i in range(len(nums)):
        num_to_compare = nums[i]
        for j in range(i + 1, len(nums)):
            compare_against = nums[j]
            if num_to_compare == compare_against:
                return True
    return False

# Solution 2: Sort and Compare
def S2_contains_duplicate(nums: List[int]) -> bool:
    nums.sort()  # O(n log(n))
    for i in range(len(nums) - 1):  # O(n)
        p1 = i
        p2 = i + 1
        if nums[p1] == nums[p2]:
            return True
    return False

# Solution 3: Set (Hash Table)
def S3_contains_duplicate(nums: List[int]) -> bool:
    hash = set()
    for num in nums:
        if num in hash:
            return True
        hash.add(num)
    return False
